### PR TITLE
ARROW-15398: [Integration-tests] Made integration tests be built in parallel and increase cache hit

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This is a draft PR to gauge interest in splitting the build of integration tests into separate stages, so that a change in a language does not trigger a rebuild in all of them.

The idea is to use docker multi-stage builds to build the integration test binaries for each language, which are then packed in the same docker image prior to running all the tests. This has 2 main consequences:

1. the build of each of the languages is done in parallel, since docker parallelizes independent stages
2. the build of a language is only done when something changed on it

Overall, the hope is to significantly reduce the time it takes to build and run the tests, as we only need to build the language that changed in a given PR, as opposed to all of them. It also makes it easier to understand what dependencies each language depends on.

There are some rough edges in the crossbow, but before handling them, I would like to gauge whether this is something that others would be interested in.

### CODE REVIEW

**Comment by pitrou** [comment]
@raulcd @assignUser Does one of you want to take a look at this?

**Comment by assignUser** [comment]
>I would like to gauge whether this is something that others would be interested in.

💯 Improving the (docker & other) caching for our builds is something I am currently looking at so this comes just at the right time :D

**Comment by assignUser** [comment]
>I would like to gauge whether this is something that others would be interested in.

💯 Improving the (docker & other) caching for our builds is something I am currently looking at so this comes just at the right time :D

File: ci/docker/conda-integration.dockerfile [hunk 1]
```diff
@@ -72,3 +70,102 @@ ENV ARROW_BUILD_INTEGRATION=ON \
...
+    mamba clean --all --force-pkgs-dirs
+
+COPY ./java /arrow/java
+COPY ./format /arrow/format
+COPY ./ci/scripts/java_build.sh /arrow/ci/scripts/java_build.sh
+
+RUN /arrow/ci/scripts/java_build.sh /arrow /build
+
+##################### RUST ######################
+FROM ${repo}:${arch}-conda-cpp AS builder-rust
```
Comment (raulcd) [inline ci/docker/conda-integration.dockerfile:120]: The only concern I have is that the today we don't require to have Rust cloned locally but I am not sure how we can skip a stage on docker or how we could make this conditional following this approach
As expected if I try to build locally I get the error for Rust:
```
------                                                                                                                                                         
 > [builder-rust 3/5] COPY ./rust /arrow/rust:                                                                                                                 
------                                                                                                                                                         
failed to solve: failed to compute cache key: "/rust" not found: not found 
```

**Review by raulcd — COMMENTED** [review]
This looks great to me. We should probably do more of these. I have one question on how we are going to solve RUST being optional.

**Pull request closed without merging**